[stdlib] Fix strideable methods for large unsigned values

natecook1000 opened the pull request:
`Int` is still an insufficient stride type for binary integers, but this improves the situation for values at the extremes of the concrete integer types.

natecook1000 committed:
[stdlib] Add more tests for strideable behavior

natecook1000 commented:
@swift-ci Please test

natecook1000 commented:
@swift-ci Please smoke benchmark

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -1626,22 +1626,33 @@ extension BinaryInteger {
+    let sameSign = !Self.isSigned || (isNegative == (other < 0))
+
+    if sameSign {
+      if self.magnitude > other.magnitude {
```
It looks like one can simplify the code a bit.

Say we calculate `delta = self.magnitude - other.magnitude`.

example | delta | delta < 0 | isNegative | finalResult
-- | -- | -- | -- | --
2.distance(to: 5) | -3 | true | false | 3 (== `-delta`)
5.distance(to: 2) | 3 | false | false | -3 (== `-delta`)
(-2).distance(to: -5) | -3 | true | true | -3 (== 'delta')
(-5).distance(to: -2) | 3 | false | true | 3 (== `delta`)

So it looks like this should work..

```swift
if sameSign {
  let delta = self.magnitude - other.magnitude
  if let result = Int(exactly: delta) {
    return isNegative ? result : -result
  }
} else {
  // this part is unchanged
  if let result = Int(exactly: self.magnitude + other.magnitude) {
    return isNegative ? result : -result
  }
}
```

which can be simplified a little more:

```swift
let distance = sameSign
  ? (self.magnitude - other.magnitude)
  : (self.magnitude + other.magnitude)
if let result = Int(exactly: distance) {
  return isNegative ? result : -result
}
```

If this is correct (which I'm not 💯 sure), this should probably lead to a better code. I would be interested in looking at the SIL and IR being generated.

moiseev replied:
Oh... I take that back, the type can be unsigned and then the negative delta will not be representable ;-(

natecook1000 replied:
Right… Between signed types not being able to represent their whole range and unsigned types not being able to represent negative values, this is kind of an annoying method.

xwu replied:
@natecook1000 I *think* there may be a slightly more simple way to do this that may also be a little more performant. Won't have time to try until Saturday, but I will update if it works out :) Anyway, glad that this is getting some attention.

moiseev reviewed (commented):
I see you decided to leave the default implementations on `BinaryInteger`. Why so? I thought we agreed that enforcing `Int` as a `Stride` for any big-integer is not the best idea ;-)

natecook1000 commented:
@swift-ci Please test

natecook1000 commented:
@swift-ci Please smoke benchmark

natecook1000 commented:
@swift-ci Please test

natecook1000 commented:
@swift-ci Please smoke test

xwu commented:
@natecook1000 I think I've got it. By inspection, I believe it eliminates one branch in most code paths as compared to your alternative, and it eliminates uses of `magnitude` along some code paths, which may be free for unsigned integers but potentially more expensive for sign-magnitude BigInt instances. I've also tweaked `advanced(by:)` so that it never performs a unary negation and a subtraction in preference to a single addition if the type is signed. See if it benchmarks well:

```swift
extension BinaryInteger {
  @inline(__always)
  public func distance(to other: Self) -> Int {
    if !Self.isSigned {
      if self > other {
        if let result = Int(exactly: self - other) { return -result }
      } else {
        if let result = Int(exactly: other - self) { return result }
      }
    } else {
      let isNegative = self < (0 as Self)
      if isNegative == (other < (0 as Self)) {
        if let result = Int(exactly: other - self) { return result }
      } else {
        if let result = Int(exactly: self.magnitude + other.magnitude) {
          return isNegative ? result : -result
        }
      }
    }
    _preconditionFailure("Distance is not representable in Int")
  }

  @_transparent
  public func advanced(by n: Int) -> Self {
    return !Self.isSigned && n < 0 ? self - Self(-n) : self + Self(n)
  }
}
```

natecook1000 commented:
@xwu That looks terrific! The only case I can see that we aren't handling well is advancing signed integers by more than their maximum (or by less than their minimum): `(-100 as Int8).advanced(by: 200)` will trap even though the result is representable. My version traps on that too, but the current implementation doesn't (though it traps other places that aren't good).

xwu commented:
@natecook1000 I hadn't considered that. We _can_ avoid such an issue, but it isn't pretty:

```swift
  public func advanced(by n: Int) -> Self {
    if let i = Int(exactly: self) {
      let (j, overflow) = i.addingReportingOverflow(n)
      if !overflow, let result = Self(exactly: j) { return result }
    }
    return !Self.isSigned && n < 0 ? self - Self(-n) : self + Self(n)
  }
```

xwu commented:
...more correctly, though:

```swift
  public func advanced(by n: Int) -> Self {
    if !Self.isSigned {
      return n < (0 as Int) ? self - Self(-n) : self + Self(n)
    }
    if (self < (0 as Self)) == (n < (0 as Self)) {
      return self + Self(n)
    }
    return self.magnitude < n.magnitude ? Self(Int(self) + n) : self + Self(n)
  }
```

xwu committed:
[stdlib] Fix strideable methods for large unsigned values

Int is still an insufficient stride type for binary integers,
but this improves the situation for values at the extremes of the
concrete integer types.

natecook1000 commented:
@swift-ci Please smoke benchmark

natecook1000 commented:
@swift-ci Please test

xwu commented:
@natecook1000 How would you interpret these benchmarks?

natecook1000 commented:
With the caveat that I'm not great at interpreting these, you can really see the effect of the branching in the unoptimized runs of the new tests—`RangeIterationSigned` uses `Int`s, so it gets straight addition and subtraction. The optimized versions either show that the optimizer does a great job of reducing that effect or that the optimizer has defeated my benchmark once again (although a nonzero time is promising). You?

xwu commented:
That sounds reasonable; I'd really love some way to compare the same benchmark before and after the fix though.

moiseev commented:
@natecook1000 if you land the benchmark in a separate PR, and then re-run the benchmarks here, we will be able to see the effect of this PR.

airspeedswift commented:
It looks like this also resolves `(2 as UInt).distance(to: 1 as UInt)` which is currently erroring when it shouldn't (and trapping if you break it into 2 expressions)

moiseev commented:
rdar://problem/34707531 for the reference

moiseev commented:
@natecook1000 I extracted the benchmarks from here into a separate PR.

natecook1000 commented:
@moiseev Thanks! Looks like the benchmark harness changed somewhat, hadn't seen that…

natecook1000 commented:
@swift-ci Please test

natecook1000 commented:
@swift-ci Please test Linux platform

moiseev commented:
OK. Now we have the benchmarks finally.
@swift-ci Please smoke benchmark

xwu commented:
Huh.

moiseev commented:
@xwu if you're referring to this weird comment above, it's the artifact of a failed compiler performance build. I think...

xwu commented:
Well then, is this good to merge?

moiseev commented:
@swift-ci Please benchmark

moiseev commented:
@swift-ci please smoke test compiler performance

natecook1000 committed:
[stdlib] Try inlining integer stride methods

moiseev commented:
@swift-ci Please benchmark

moiseev commented:
@swift-ci please smoke test compiler performance

moiseev commented:
@swift-ci Please test

moiseev commented:
Just in case... The benchmarks build is actually running: https://ci.swift.org/job/swift-PR-osx-perf/37/
